crates/core/src/host/instance_env.rs

```diff
@@ -426,9 +426,115 @@ impl InstanceEnv {
+    iter: I,
+    buf: Vec<u8>,
+    state: BSatnCompactorState,
+}
+
+impl<I: Iterator<Item = Result<IterItem, NodesError>>> BSatnCompactor<I> {
+    fn new(iter: I) -> Self {
+        BSatnCompactor {
+            iter,
+            buf: Vec::new(),
+            state: BSatnCompactorState::Start,
+        }
+    }
+}
+
+impl<I: Iterator<Item = Result<IterItem, NodesError>>> Iterator for BSatnCompactor<I> {
+    type Item = Result<Vec<u8>, NodesError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // For now, just send buffers over a certain fixed size.
+        fn should_yield_buf(buf: &Vec<u8>) -> bool {
+            const SIZE: usize = 64 * 1024;
+            buf.len() >= SIZE
+        }
+
+        let BSatnCompactor {
+            ref mut iter,
+            ref mut buf,
+            ref mut state,
+        } = self;
+
+        if let BSatnCompactorState::Complete = state {
+            return None;
+        }
+
+        loop {
+            match (*state, iter.next()) {
+                (BSatnCompactorState::Start, Some(Ok(IterItem::Header(header)))) => {
+                    *state = BSatnCompactorState::Iterating;
+                    header.encode(buf);
+                    let mut next = Vec::new();
+                    std::mem::swap(&mut next, buf);
+
+                    // header gets a buffer by itself
+                    return Some(Ok(next));
+                }
+                (BSatnCompactorState::Iterating, Some(Ok(IterItem::Row(row)))) => {
+                    row.encode(buf);
+                    if should_yield_buf(&*buf) {
+                        let mut next = Vec::new();
+                        std::mem::swap(&mut next, buf);
+                        return Some(Ok(next));
+                    }
+                }
+                (BSatnCompactorState::Iterating, None) => {
+                    *state = BSatnCompactorState::Complete;
+                    if !buf.is_empty() {
+                        // empty vecs don't allocate :^)
+                        let mut next = Vec::new();
+                        std::mem::swap(&mut next, buf);
+                        return Some(Ok(next));
+                    } else {
+                        return None;
+                    }
+                }
+                (_, Some(Err(err))) => {
+                    *state = BSatnCompactorState::Complete;
+                    return Some(Err(err));
+                }
+                (BSatnCompactorState::Start, _) => panic!(
+                    "{}: Wrapped iterator did not produce a starting header!",
+                    std::any::type_name::<Self>()
+                ),
+                (BSatnCompactorState::Iterating, Some(Ok(IterItem::Header(_)))) => panic!(
+                    "{}: Wrapped iterator produced too many headers!",
+                    std::any::type_name::<Self>()
+                ),
+                (BSatnCompactorState::Complete, _) => unreachable!(),
```

Comment:
I think there's enough states in this state machine that I'd appreciate a brief comment about why this is unreachable. Probably related to the small comments I'm asking for on the `BSatnCompactorState` arms.